packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart

```diff
@@ -0,0 +1,220 @@
+import 'package:amplify_api/src/graphql/web_socket/blocs/web_socket_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/state/web_socket_state.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/web_socket_types.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/ws_subscriptions_event.dart';
+import 'package:amplify_api/src/graphql/web_socket/web_socket_message_stream_transformer.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:async/async.dart';
+import 'package:flutter/material.dart';
+import 'package:stream_transform/stream_transform.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+/// {@template api.web_socket_service}
+/// Internal stateless [WebSocketService] abstract.
+/// {@endtemplate}
+abstract class WebSocketService implements Closeable {
+  /// Used to initialize a [WebSocketChannel] connection and return transformed
+  /// messages as [WebSocketEvent]
+  Stream<WebSocketEvent> init(
+    WebSocketState state,
+  );
+
+  /// Subscribe a [GraphQLRequest] inside the provided [WsSubscriptionBloc] to a open [WebSocketChannel]
+  Future<void> register(
+    ConnectedState state,
+    GraphQLRequest<Object?> request,
+  );
+
+  /// Unsubscribe a [GraphQLRequest] inside the provided [WsSubscriptionBloc]
+  Future<void> unsubscribe(
+    String subscriptionId,
+  );
+
+  /// Reconnect a [WebSocketChannel]
+  Future<void> reconnect();
+}
+
+/// {@template api.web_socket_service}
+/// Internal stateless [WebSocketService] implementation.
+/// {@endtemplate}
+class AmplifyWebSocketService
+    with AWSDebuggable, AmplifyLoggerMixin
+    implements WebSocketService, Closeable {
+  /// Service that contains the business logic for subscription management
+  AmplifyWebSocketService();
+
+  /// Sink used for sending messages to AppSync
+  @visibleForTesting
+  WebSocketSink? sink;
+
+  @override
+  Stream<WebSocketEvent> init(
+    WebSocketState state,
+  ) {
+    final sc = StreamCompleter<WebSocketEvent>();
+    _init(state, sc).catchError(sc.setError);
+
+    return sc.stream;
+  }
+
+  Future<void> _init(
+    WebSocketState state,
+    StreamCompleter<WebSocketEvent> sc,
+  ) async {
+    assert(sink == null,
+        'Connection already exists, recreate service or clear sink');
+
+    try {
+      const webSocketProtocols = ['graphql-ws'];
+      final connectionUri = await generateConnectionUri(
+        state.config,
+        state.authProviderRepo,
+      );
+
+      final channel = WebSocketChannel.connect(
+        connectionUri,
+        protocols: webSocketProtocols,
+      );
+      sink = channel.sink;
+
+      final subscriptionStream = transformStream(channel.stream);
+
+      _send(WebSocketConnectionInitMessage());
+
+      return sc.setSourceStream(subscriptionStream);
+    } on Object catch (e, st) {
+      logger.error("error in init!", e, st);
```

Comment:
Need to rethrow here so that `sc.setError` gets called above

```suggestion
      logger.error("error in init!", e, st);
      rethrow;
```